client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -51,103 +52,84 @@
     }
 
     /**
-     * Triggers a new sniffing round and explicitly takes out the failed host provided as argument
+     * Triggers a new immediate sniffing round, which will schedule a new round in sniffAfterFailureDelayMillis ms
      */
-    public void sniffOnFailure(HttpHost failedHost) {
-        this.task.sniffOnFailure(failedHost);
+    public final void sniffOnFailure() {
+        scheduleNextRound(0L, sniffAfterFailureDelayMillis, true);
     }
 
-    @Override
-    public void close() throws IOException {
-        task.shutdown();
+    private void scheduleNextRound(long delay, long nextDelay, boolean mustCancelNextRound) {
+        Task task = new Task(nextDelay);
+        Future<?> nextFuture = task.schedule(delay);
+        Future<?> previousFuture = nextTask.getAndSet(nextFuture);
+        if (mustCancelNextRound) {
+            previousFuture.cancel(false);
+        }
     }
 
-    private static class Task implements Runnable {
-        private final HostsSniffer hostsSniffer;
-        private final RestClient restClient;
-
-        private final long sniffIntervalMillis;
-        private final long sniffAfterFailureDelayMillis;
-        private final ScheduledExecutorService scheduledExecutorService;
-        private final AtomicBoolean running = new AtomicBoolean(false);
-        private ScheduledFuture<?> scheduledFuture;
-
-        private Task(HostsSniffer hostsSniffer, RestClient restClient, long sniffIntervalMillis, long sniffAfterFailureDelayMillis) {
-            this.hostsSniffer = hostsSniffer;
-            this.restClient = restClient;
-            this.sniffIntervalMillis = sniffIntervalMillis;
-            this.sniffAfterFailureDelayMillis = sniffAfterFailureDelayMillis;
-            SnifferThreadFactory threadFactory = new SnifferThreadFactory(SNIFFER_THREAD_NAME);
-            this.scheduledExecutorService = Executors.newScheduledThreadPool(1, threadFactory);
-            scheduleNextRun(0);
+    final class Task implements Runnable {
+        final long nextTaskDelay;
+
+        Task(long nextTaskDelay) {
+            this.nextTaskDelay = nextTaskDelay;
         }
 
-        synchronized void scheduleNextRun(long delayMillis) {
-            if (scheduledExecutorService.isShutdown() == false) {
-                try {
-                    if (scheduledFuture != null) {
-                        //regardless of when the next sniff is scheduled, cancel it and schedule a new one with updated delay
-                        this.scheduledFuture.cancel(false);
-                    }
-                    logger.debug("scheduling next sniff in " + delayMillis + " ms");
-                    this.scheduledFuture = this.scheduledExecutorService.schedule(this, delayMillis, TimeUnit.MILLISECONDS);
-                } catch(Exception e) {
-                    logger.error("error while scheduling next sniffer task", e);
-                }
-            }
+        Future<?> schedule(long delay) {
```

Comment:
Maybe drop this method and call `scheduler.schedule(theTask, delay)` all the places?

Reply:
makes sense